feat: Auto research skill (#2419)
NVIDIA-NeMo/RL: skills/auto-research/SKILL.md

---
name: auto-research

> description: Autonomous NeMo RL experiment driver for accuracy-focused recipe search in this repository. Use when coding agents should inspect recipe inheritance and backend code paths, run repeatable experiments, and record each hypothesis on its own git branch under a shared prefix with a TSV experiment log.

Reviewer: [`skills/auto_research/SKILL.md:3`](https://github.com/NVIDIA-NeMo/RL/blob/ca90191aa2e6d16ba08d5e6010ff409da356d4e1/skills/auto_research/SKILL.md#L3)

Every other skill in `skills/` (11 out of 11) includes a `when_to_use` frontmatter field — the system uses it to decide when to suggest the skill. Without it, this skill won't be auto-surfaced during skill discovery. Consider adding one:

```suggestion
description: Autonomous NeMo RL experiment driver for accuracy-focused recipe search in this repository. Use when coding agents should inspect recipe inheritance and backend code paths, run repeatable experiments, and record each hypothesis on its own git branch under a shared prefix with a TSV experiment log.
when_to_use: Running autonomous experiment campaigns; iterating on recipe accuracy; 'auto research', 'experiment loop', 'hypothesis testing', 'accuracy search'.
```

Author: fixed

> description: Autonomous NeMo RL experiment driver for accuracy-focused recipe search in this repository. Use when coding agents should inspect recipe inheritance and backend code paths, run repeatable experiments, and record each hypothesis on its own git branch under a shared prefix with a TSV experiment log.

Author: fixed

---

# Auto Research

Run iterative NeMo RL experiments in this repository with accuracy as the objective and git as the ledger.

Treat dependencies as ready, but choose the runtime deliberately. Use the recipe's authoritative metric as the source of truth. Keep changes small, reproducible, and simple. Preserve unrelated user work.

## Workflow

1. Inspect the current git state and identify unrelated user changes before branching.
2. Use a shared branch prefix. Prefer a user-provided one; otherwise create a suggestive default such as `autoresearch/2026-03-24-dapo-qwen2p5`.
3. Read the target recipe, its parents, and the relevant code paths in `examples/run_grpo.py`, `nemo_rl/models/`, `nemo_rl/algorithms/`, and `docs/`.
4. Translate any user stop rule into explicit values you can monitor, such as `target_experiments`, `campaign_deadline`, `per_experiment_timeout`, or `target_metric`.
5. Verify required data, checkpoints, runtime inputs, and the launcher.
6. Create an untracked TSV log and per-experiment log directory.
7. Run a baseline first on `<prefix>/baseline` if none exists.

> For GPU, CPU-heavy, distributed, or long-running work, use the relevant runtime skill before launching: `slurm` on a Slurm head node or `launch-nemo-rl` for nrl-k8s/Kubernetes. Run locally only for light inspection, dry runs, and short non-GPU checks.

Reviewer: [`skills/auto_research/SKILL.md:22`](https://github.com/NVIDIA-NeMo/RL/blob/ca90191aa2e6d16ba08d5e6010ff409da356d4e1/skills/auto_research/SKILL.md#L22)

This references a `slurm` runtime skill, but no such skill exists in `skills/`. The only cluster-launch skill is `launch-nemo-rl` (for nrl-k8s/Kubernetes). Consider clarifying:

```suggestion
For GPU, CPU-heavy, distributed, or long-running work, use the relevant runtime skill before launching (e.g., `launch-nemo-rl` for nrl-k8s/Kubernetes, or your environment's native launcher for Slurm). Run locally only for light inspection, dry runs, and short non-GPU checks.
```

Author: fixed

## Branching

- Put every experiment on its own branch under the shared prefix.

> - Keep every branch, even for failed or weak ideas.

Author: fixed

- Never stash, reset, or overwrite unrelated user changes silently. If dirty files overlap the experiment, use a separate worktree or ask before proceeding.

See `references/git-workflow.md` for the exact pattern.

## Loop

1. Pick one concrete hypothesis.
2. Create a branch such as `autoresearch/2026-03-24-dapo-qwen2p5/prompt-compact-schema`.
3. Edit the smallest set of files needed.
4. Commit the hypothesis.
5. Before launching the run, check the monitored stop conditions. Do not stop early unless one is already clearly met.
6. Identify the authoritative metric source from the recipe or logging code, then run with a unique log path:

```bash
LOG_DIR=reports/auto_research/<campaign>/<experiment>
mkdir -p "$LOG_DIR"
uv run <entrypoint> > "$LOG_DIR/run.log" 2>&1
```

7. If the user gave a per-experiment wall-clock limit, enforce it explicitly. Prefer a recipe-level timeout when one already exists; otherwise wrap the command with an external timeout. If both exist, honor the tighter limit.
8. Extract the primary metric with a command appropriate for the actual log format. If extraction is empty, inspect the last log lines and the recipe's logging path before marking the run.

> 9. Record branch, parent commit, commit, recipe, metric name, metric value, memory, elapsed time, launcher or job id, command, log path, status, and description in the TSV, along with enough timing or count information to evaluate the stop rule.

Reviewer: [`skills/auto_research/SKILL.md:51`](https://github.com/NVIDIA-NeMo/RL/blob/ca90191aa2e6d16ba08d5e6010ff409da356d4e1/skills/auto_research/SKILL.md#L51)

A few inconsistencies between this field list and the TSV template in `references/experiment-log-template.md`:

1. **Missing `index`**: The template's first column is `index` (the attempted experiment count), which is critical for stop-condition logic (lines 56, 63). This step doesn't mention recording it.
2. **Unit ambiguity**: This says "memory" and "elapsed time", but the template uses `memory_gb` and `elapsed_min`. An agent following this step may record the wrong units.
3. **Field split**: "launcher or job id" reads as one field, but the template has two separate columns (`launcher`, `job_id`).

Consider updating this line to match the template column names exactly, e.g.:
```suggestion
9. Record index, branch, parent commit, commit, recipe, metric name, metric value, memory (GB), elapsed time (minutes), launcher, job id, command, log path, status, and description in the TSV, along with enough timing or count information to evaluate the stop rule.
```

Author: fixed
